### PR TITLE
Fixed wonkey logic in TypedRangeStrideSegment, added barrage of unit tests

### PR DESCRIPTION
**Pull request by ajkunen**
The TypedRangeStrideSegment had some implementation issues that this PR aims to resolve:
1) didn't handle non-exact ending values correctly, such as  TypedRangeStrideSegment(1, 13, 3)
2) didn't handle negative strides at all, such as TypedRangeStrideSegment(10, 3, -1)

I've gotten this all working, at least sequentially. Still need to make sure it works with OMP/CUDA/etc.

### CODE REVIEW

File: include/RAJA/index/RangeSegment.hpp [hunk 1]
```diff
@@ -258,10 +279,13 @@ struct TypedRangeStrideSegment {
                                            StorageT stride)
       : m_begin(iterator(begin, stride)),
         m_end(iterator(end, stride)),
-        m_size((end - begin) >= stride
-                   ? (end - begin) / stride + ((end - begin) % stride)
-                   : 0)
+        // essentially a ceil((end-begin)/stride) but using integer math,
+        // and allowing for negative strides
+        m_size((end - begin + stride - ( stride > 0 ? 1 : -1  ) ) / stride)
```
Comment (willkill07) [inline include/RAJA/index/RangeSegment.hpp]: This is much cleaner

**Comment by willkill07** [comment]
CI is complaining about signed/unsigned comparison:

```cpp
for (size_t i = 0; i < segment.size(); ++i)
```

My hack of always making this work:
```cpp
using IterType = typename std::remove_const<decltype(segment.size())>::type;
for (IterType i = 0; i < segment.size(); ++i)
```

File: test/unit/cpu/test-segments.cpp [hunk 2]
```diff
@@ -150,7 +150,7 @@ TEST(RangeStrideSegmentTest, values_forward_stride1)
   
   ASSERT_EQ(segment.size(), 6);
  
-  for(size_t i = 0;i < segment.size();++ i){
+  for(RAJA::Index_type i = 0;i < segment.size();++ i){
```
Comment (rhornung67) [inline test/unit/cpu/test-segments.cpp]: This is really wild use of blank spaces. Do you always code like this? :-)

File: include/RAJA/index/RangeSegment.hpp [hunk 3]
```diff
@@ -200,18 +200,31 @@ struct TypedRangeSegment {
...
+ * \brief  Segment class representing a strided and typed range of indices
  *
  * \tparam StorageT the underlying data type for the Segment
  *
  * A TypedRangeStrideSegment models an Iterable interface:
  *
  *  begin() -- returns an iterator (TypedRangeStrideSegment::iterator)
  *  end() -- returns an iterator (TypedRangeStrideSegment::iterator)
- *  size() -- returns the total size of the Segment
+ *  size() -- returns the total iteration size of the Segment, not
```
Comment (rhornung67) [inline include/RAJA/index/RangeSegment.hpp]: 👍

**Pull request merged**